Std/Tactic/Pattern/Utils.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2023 Anand Rao Tadipatri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anand Rao Tadipatri
+-/
+import Lean.Elab.Term
+import Lean.Elab.Tactic
+import Lean.SubExpr
+import Lean.Meta.ExprLens
+import Lean.Meta.KAbstract
+import Lean.HeadIndex
+
+open Lean Meta Elab Tactic
+
+/-!
+
+Basic utilities for tactics that target goal locations through patterns and their occurrences.
+
+The code here include:
+- Functions for expanding syntax for patterns and occurrences into their corresponding expressions
+- Code for generating and finding the occurrences of patterns in expressions
+
+The idea of referring to sub-expressions via patterns and occurrences is due to Yaël Dillies.
+
+-/
+open Parser.Tactic Conv
+
+/-- Refer to a set of subexpression by specifying a pattern.
+
+For example, if hypothesis `h` says that `1 + (2 + 3) = 1 + (2 + 3)`, then
+`(occs := 2 3) _ + _ at h` refers to the two expression `2 + 3`,
+because it first skips `1 + (2 + 3)`, and matches with `2 + 3`,
+which instantiates the pattern to be `2 + 3`, so the next match is the
+second instance of `2 + 3`. -/
```

Comment:
It is rather unhelpful here that you use `2` and `3` here to mean different things! This would be immensely clearer if you used `17 + (37 + 19)` or something.

Also, the phrase "the two expression `2 + 3`" doesn't make sense to me. Do you mean "the two occurrences of the expression `2 + 3`"?